### PR TITLE
Lf 4078 update v 2 table component to support animals table

### PR DESCRIPTION
**Pull request by Duncan-Brain**
**Description**
Main thing I think I struggled with here was the complex styling and the types (made a request for help inline). 

Table body border styling I failed to complete:
Screenshot 2024-02-22 at 12 55 03 AM

Screenshot 2024-02-22 at 1 41 46 AM

A transparent border with border stroke of 16px is close but hides the header bottom border. So I settled for just a gap between head and body.

Included (from ticket):
- Replace the old sort icon with new ones. The new icons should look good on the existing finance tables.
  - Updated the width to accomodate new sort icon
- Columns should have the option to be sortable or unsortable.
- Decide if the alternate colour should be implemented in the table component and implement it if necessary.
- Create a story and make sure it supports animal tables. (chose not to do animal tables just the new component variations)

Also included but not explicitly scoped in ticket:
- Table component needed animals table cell types available (hover pill, icon, chevron)
- Basic implementation on the route to showcase/view

Not included:
- Implement “Load more” button properly. Removed from scope as it is superficial since we load everything anyways discussed during grooming.
  - I tried Sayakas approach to set minRows=data.length but it needed a useEffect to handle loading data
- Checkboxes are not in scope.
- Cell types we don't have implemented yet like animal locations, custom animal images
- [LF-4103](https://lite-farm.atlassian.net/browse/LF-4103) should hopefully include:
  - Returning text names of types and breeds
  - Cleaning data to support correct sorting (sorting is based on one single data id and some columns are `default_id || custom_id`)
  - Fixing width of columns to be more appropriate
  -  other implementation details

Jira link: [LF-4078](https://lite-farm.atlassian.net/browse/LF-4078)

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

[LF-4103]: https://lite-farm.atlassian.net/browse/LF-4103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [inline packages/webapp/src/components/Table/types.ts:92]: Really struggled to get this right I feel like ReactNode should work but doesn't.

File: packages/webapp/src/components/Table/Table.tsx
Comment (Duncan-Brain) [inline packages/webapp/src/components/Table/Table.tsx:31]: Cheating using the any type here couldn't get props to reduce scope here.

File: packages/webapp/src/components/Table/types.ts
Comment (kathyavini) [reply packages/webapp/src/components/Table/types.ts:92]: This gives you a ts error right in the file?

```ts
export type FunctionReturnsComponent = (props: any) => ReactNode;
```

Or you mean it gives you an error when you use it in context? If the second, I think it's because an icon isn't a function that returns a ReactNode, it is just a ReactNode:

```ts
export type IconTextProps = {
  text?: string | null;
  icon: ReactNode;
  subtext?: string | null;
};
```

File: packages/webapp/src/components/Table/Table.tsx
Comment (kathyavini) [reply packages/webapp/src/components/Table/Table.tsx:31]: This is kind of a gnarly component to be typing with the key value pair mapping and the double wrapping!!

If you don't mind doing the key value pair mapping right in the component itself (instead of with an intermediary constant), you could write a discriminated union for the props and get the type checking on the component when you call it based on what you provide for `kind`):

```ts
type ReactTable = ReactTableProps & { kind: TableType.V1 };
type MuiTable = MuiTableProps & { kind: TableType.V2 };

type TableProps = ReactTable | MuiTable;

export default function Table({ kind, ...props }: TableProps) {
  switch (kind) {
    case TableType.V1:
      return <ReactTable {...(props as ReactTableProps)} />;
    case TableType.V2:
      return <MuiTable {...(props as MuiTableProps)} />;
  }
}

```

Would that work for you? I don't know how to keep `KindComponentKVP` as a separate object and still keep the narrowing.

**Review by kathyavini — COMMENTED** [review]
It looks really good! And there it is, actually working in inventory 😁 

The only styling thing that troubled me was the way that the column labels shift a little when you apply sort (I made a note inline below). Aaaaaaand (zero percent an issue with your code) buuuuut... I'm not sure I like the new icons on the finance tables! I get the striving for consistency, but.... the old chevrons just fit the vibe there more!!

And my good job... it's ALL in TypeScript!! 👏 🎉  You had some pretty tough components to type (like your enum approach to tables + cells, which I liked) so I bet you had more than your fair share of struggling with the type checker!

I made a few notes for TS stuff that was a little different than how I would have typed; some of it might merit discussion at Tech Daily so we can settle on codebase patterns.

File: packages/webapp/src/routes/AnimalsRoutes.jsx
Comment (Duncan-Brain) [reply packages/webapp/src/routes/AnimalsRoutes.jsx:23]: Yeah, it was but I think it was from having both index.jsx and index.tsx when I merged integration. I thiiink I can remove this if I restart vscode thanks for the reminder!

File: packages/webapp/src/components/Table/Cell/CellTypes/HoverPillOverflow.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Cell/CellTypes/HoverPillOverflow.tsx:24]: Hmm true, yeah for some reason I didn't forsee having to make all the Cell types components too (but also seemed necessary for the ticket ask) so I probably skimped on attention to them a bit.

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts:93]: I am having trouble keeping up with the flakiness of typescript haha. 

I though I was seeing other errors too in /V2 etc. Will go this route thanks!

File: packages/webapp/src/components/Table/Table.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Table.tsx:31]: Nice thank you! I am still only like 1 or 2 hrs in on my TS course. 

Okay -- I think I really like this pattern and intend to use it moving forward since it is quite effective for passing along the typescript typings. There only seems to *maybe* be a performance hit for large numbers of switch statements vs lookup table (which I don't think would happen in our cases) (https://stackoverflow.com/questions/37730199/switch-vs-object-lookup-performance-since-jsperf-is-down, https://stackoverflow.com/questions/11617091/in-a-switch-vs-dictionary-for-a-value-of-func-which-is-faster-and-why) 

Thanks for sharing the keyword the [discriminated union docs ](https://basarat.gitbook.io/typescript/type-system/discriminated-unions) are really helpful, I also looked into [Strategy Patterns](https://dev.to/carlillo/design-patterns---strategy-pattern-in-javascript-2hg3) this morning to see if there was anything recommended or popular -- but didn't come up with much.

File: packages/webapp/src/components/Table/Table.tsx
Comment (antsgar) [inline packages/webapp/src/components/Table/Table.tsx:18]: I think we should import these as `TableV1` and `TableV2` -- the implementation details of which library we used to build them should be transparent to the rest of the components. It would also be nice for clarity to move the files related to v1 to their own v1 folder, similarly to v2.

File: packages/webapp/src/components/Animals/Inventory/index.tsx
Comment (antsgar) [inline packages/webapp/src/components/Animals/Inventory/index.tsx:44]: I think we'll need to pass in `dense={false}` here so that the row height is 56px instead of 40px

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (antsgar) [inline packages/webapp/src/components/Table/Cell/index.tsx:23]: The name of this type is confusing to me tbh 😅 what does KVP stand for?

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (antsgar) [inline packages/webapp/src/components/Table/Cell/index.tsx:43]: I would use `cellType` or `type` as the variable name instead of `kind` everywhere to make it easier to understand that it's an option out of the CellType enum

**Review by antsgar — COMMENTED** [review]
Left a few comments but it's looking great, nice to see it start to come together! 🙌

File: packages/webapp/src/components/Animals/Inventory/index.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Animals/Inventory/index.tsx:27]: I think that is more of a [LF-4103](https://lite-farm.atlassian.net/browse/LF-4103) thing. The Animal Inventory component is just a convenience right now.

File: packages/webapp/src/stories/Table/TableCells.stories.jsx
Comment (Duncan-Brain) [reply packages/webapp/src/stories/Table/TableCells.stories.jsx:37]: Thanks I forgot to update the naming convention from disabled > sortable!

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts:1]: I think I like that too. For some things like enums I am essentially using it as a replacement for constants.js.

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts:54]: Yeah, thats what I wanted FunctionReturnsComponent to be but I had those type issues. If you could check again once I add the changes.

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts:71]: Cool thanks!

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts:79]: Thanks added number !

File: packages/webapp/public/locales/en/translation.json
Comment (Duncan-Brain) [reply packages/webapp/public/locales/en/translation.json:158]: I personally have a feeling these are generic enough to reuse these throughout animals. Can I lowercase them and use the tolocaleUppercase() type of thing instead?

File: packages/webapp/src/containers/Animals/Inventory/index.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/containers/Animals/Inventory/index.tsx:38]: To be honest I just copied other implementations and didn't think about it. Labour table uses your suggested approach inside a useMemo() .. I can do that instead?

File: packages/webapp/src/components/Table/v2/index.jsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/v2/index.jsx:104]: Yeah... that is why. I looked into that but didn't see it available in the existing rtk stuff. I can remove this and troubleshoot isLoaded in [LF-4103](https://lite-farm.atlassian.net/browse/LF-4103) if thats ok.

File: packages/webapp/src/components/Table/Table.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Table.tsx:16]: I can rename.. I thought it might be a whole thing to edit the table structures. 

In my opinion I would like to change `Table/index.jsx` to `Table/TableV1.jsx` and `Table/Tables.tsx` to `Table/index.tsx` and move `Table/v2/index` and `Table/v2/styles` out of v2 folder to top level. With Headers getting its own folder too for Table/Headers/TableHead.jsx.

File: packages/webapp/src/components/Animals/Inventory/index.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Animals/Inventory/index.tsx:44]: Sure I can toss that in here -- I made no attempt to get styles exactly right here since I thought [LF-4103](https://lite-farm.atlassian.net/browse/LF-4103) was supposed to handle implementing what the Table component can provide.

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Cell/index.tsx]: Sorry, its gone now anyways in terms of @kathyavini Typing approach. But "Key-Value Pair" was the intent.

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Cell/index.tsx]: Maybe I should change cellType to cellKind? ...type is a reserved word in typescript haha I did do that originally lol

File: packages/webapp/public/locales/en/translation.json
Comment (antsgar) [reply packages/webapp/public/locales/en/translation.json:158]: Yeah I'm good with that, and as we keep building we can see if we actually reuse them and rename them if we don't

File: packages/webapp/src/containers/Animals/Inventory/index.tsx
Comment (antsgar) [reply packages/webapp/src/containers/Animals/Inventory/index.tsx:38]: Yup, that seems better to me!

File: packages/webapp/src/components/Table/v2/index.jsx
Comment (antsgar) [reply packages/webapp/src/components/Table/v2/index.jsx:104]: Sounds good!

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (antsgar) [reply packages/webapp/src/components/Table/Cell/index.tsx]: It's fine to use type within the context of a longer name variable like `cellType` IMO, types are PascalCase anyway so they're easily recognizable. But I'm alright with `cellKind` too if you prefer that one!

Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Cell/index.tsx]: The benefit to using keeping kind is that I can use the generic "StrategyProps" for all of them. If I switch to cellType then I need to make a new CellStrategyProps type and same for table.

File: packages/webapp/src/components/Table/v2/TableHead.jsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/v2/TableHead.jsx:55]: Yeah that looks better good point!

**Comment by Duncan-Brain** [comment]
@kathyavini Thanks for the Typings help! .. I gotta get going on that class haha 

As for the consistent header sort icons .. yeah I think I see your point it was a little more subtle before and I liked it before too but not crazy obtrusive now either 🤷 . We could ask Loic about it (after its merged is easier to show I think) and do the same destructuring of headers like cells etc. But I dunno seemed like Sayaka asked about it on the ticket so I am good to move it through.

**Review by kathyavini — COMMENTED** [review]
The functional fixes look good, which means everything I'm mentioning below is ENTIRELY about the types (AKA code that has absolute impact at runtime 😆)

I'm mentioning them because I think you want your discriminated unions to work, and they don't QUITE work yet. But @antsgar might also decide we don't care enough about ideal type checking, in which case this has my approval for runtime behaviour 👍

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/Cell/index.tsx]: Thanks yup I think I get it now!

As for PLAIN - I think I wanted a fallback but I am also down for really strict kinds. 

I used the exhaustive check style shown here: https://basarat.gitbook.io/typescript/type-system/discriminated-unions

But added the return null to keep it from returning undefined.

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts]: I think they belong in their respective tables files since they are equivalent of "PropTypes" (not prototype haha ), ...but those files are in .jsx right now and I had to stop somewhere. 

So my preference after that was a shared component `types.ts` file instead of "some other file that primarily uses it".

File: packages/webapp/src/components/Table/types.ts
Comment (Duncan-Brain) [reply packages/webapp/src/components/Table/types.ts:58]: Yeah I guess my thought process was that more specific versions of these would be helpful in the future so:

For CallbackFn I personally think it more human readable than () => void. I was thinking if we had something like `CallBackFnNavigatesToItem` -- or `CallbackFnNavigatesHome` that specific callback functions into form elements like "onGoBack" would be useful for making consistent forms (obviously not related here but generally where my head was at that defining some things is good.

For FunctionReturnsComponent I do like more human readable function expectations.  StrategyProps I have deleted for now but seemed reasonable if we were to define what a strategy pattern looks like here so for one attribute `kind: string` thats pretty obvious yah, but if we needed to add `modifier` or `containerClassname` to strategy.. yeah 🤷

File: packages/webapp/src/components/Table/Cell/index.tsx
Comment (kathyavini) [reply packages/webapp/src/components/Table/Cell/index.tsx]: I've never seen that before; neat!

**Review by kathyavini — APPROVED** [review]
Looks good, even all the non-blocking TS stuff!

**Review by antsgar — APPROVED** [review]

**Pull request merged**
